p2p/src/simulated/bandwidth.rs

```diff
@@ -27,275 +30,363 @@ pub enum Rate {
-                .members
-                .iter()
-                .filter(|&&flow_idx| unfrozen[flow_idx])
-                .count() as u128;
-            if users == 0 {
-                continue;
+            // Only track ingress when the recipient actually needs to receive the bytes.
+            if flow.delivered {
+                // Register the flow with its ingress resource if the recipient is bandwidth-limited.
+                if let Some(resource_idx) = self.constrain(
+                    Constraint::Ingress(flow.recipient.clone()),
+                    ingress_limit(&flow.recipient),
+                ) {
+                    self.attach(resource_idx, flow_idx, &mut state);
+                }
             }
 
-            // If the resource is exhausted, freeze all flows depending on it.
-            if remaining[res_idx].is_zero() {
-                limiting.clear();
-                limiting.push(res_idx);
-                min_delta = Some(Ratio::zero());
-                break;
+            // The flow participates in progressive filling until one of its constraints saturates.
+            if state.limited {
+                state.active = true;
+                self.active += 1;
             }
+            self.states.push(state);
+        }
+    }
 
-            // Potential additional rate every user could receive before hitting this limit.
-            let delta = remaining[res_idx].div_int(users);
-            match &min_delta {
-                None => {
-                    min_delta = Some(delta);
+    /// Finalize all flows that rely on a saturated resource.
+    fn finalize_all(&mut self, res_idx: usize) {
+        let members = self.resources[res_idx].members.clone();
+        for flow_idx in members {
+            self.finalize(flow_idx);
+        }
+    }
+
+    /// Finalize the rate for `flow_idx` and update every referenced resource.
+    ///
+    /// The flow's share at the moment of freezing becomes its permanent rate; afterwards we
+    /// subtract it from every referenced resource so the next progressive-filling iteration only
+    /// considers the remaining active flows.
+    fn finalize(&mut self, flow_idx: usize) {
+        let state = &mut self.states[flow_idx];
+        if !state.active {
+            return;
+        }
+
+        // The flow's max-min allocation equals the current fill level.
+        self.rates[flow_idx] = Some(self.fill.clone());
+        state.active = false;
+        self.active -= 1;
+
+        for &res_idx in &state.resources {
+            let resource = &mut self.resources[res_idx];
+            if resource.active > 0 {
+                // Stop counting the flow toward future shares once it is frozen.
+                resource.active -= 1;
+            }
+        }
+    }
+
+    /// Run the progressive filling algorithm until every constrained flow is frozen.
+    fn fill(&mut self) {
+        if self.active == 0 {
+            return;
+        }
+
```

Comment:
nit: not really needed since the `while` wont trigger and you'll just return anyway